### PR TITLE
allow array in pkg#source

### LINKED ISSUE #5387: An array `package.json#source` should build these specified entries.
Running `parcel build my-package/` should build `a.html, b.html` if `my-package/package.json` contains:
```json
{
  "source": ["a.html", "b.html"]
}
```

So be equivalent to `cd my-package && parcel build a.html b.html`.

The docs already say this should work: https://v2.parceljs.org/configuration/package-json/#source

Relevant code:
https://github.com/parcel-bundler/parcel/blob/064ba0a166ea63ccd2b623dc11324d4cd7ffb526/packages/core/core/src/requests/EntryRequest.js#L98-L126

### PR DESCRIPTION
**Pull request by DeMoorJasper**
# ↪️ Pull Request

Running parcel build my-package/ should build a.html, b.html if my-package/package.json contains:

```json
{ "source": ["a.html", "b.html"] }
```

credits to @AceTheCreator for creating the original PR

Closes #5387
Closes #5494

TODO:
- [ ] Write a test

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.86s | +65.00ms |
| Cached | 397.00ms | +10.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 100.00ms | +28.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 99.00ms | -13.00ms 🚀 |
| dist/modern/index.9e1f1458.js | 1.11kb | +0.00b | 585.00ms | -333.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 576.00ms | +86.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 69.00ms | +20.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 41.00ms | -8.00ms 🚀 |
| dist/legacy/index.33d06174.js | 1.11kb | +0.00b | 74.00ms | +9.00ms ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 77.00ms | +12.00ms ⚠️ |
| dist/legacy/index.44edca46.css | 77.00b | +0.00b | 73.00ms | +8.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 17.03s | +194.00ms |
| Cached | 875.00ms | +66.00ms ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/NotFound.9ae6337a.js | 538.00b | +0.00b | 215.00ms | -36.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 483.12kb | +0.00b | 108.00ms | +11.00ms ⚠️ |
| dist/PermalinkedComment.bb8e38c2.js | 4.24kb | +0.00b | 108.00ms | +12.00ms ⚠️ |
| dist/UserProfile.fc71f0f5.js | 1.78kb | +0.00b | 108.00ms | +12.00ms ⚠️ |
| dist/NotFound.9ae6337a.js | 538.00b | +0.00b | 92.00ms | +8.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 77.00ms | +9.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.72m | -1.29s |
| Cached | 2.50s | +33.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/EmojiPickerComponent.5f979d38.js | 139.77kb | +0.00b | 10.15s | +1.52s ⚠️ |
| dist/Toolbar.022bc487.js | 100.20kb | +0.00b | 13.78s | +862.00ms ⚠️ |
| dist/media-viewer.0266bd59.js | 73.48kb | +0.00b | 10.42s | +1.74s ⚠️ |
| dist/card.2551d1c7.js | 54.67kb | +0.00b | 8.86s | +1.66s ⚠️ |
| dist/card.30bd2aa3.js | 52.44kb | +0.00b | 11.22s | +1.06s ⚠️ |
| dist/Modal.ed4344dc.js | 39.59kb | +0.00b | 3.30s | -214.00ms 🚀 |
| dist/component.b15bf9ac.js | 31.08kb | +0.00b | 3.44s | +206.00ms ⚠️ |
| dist/esm.4af22478.js | 27.87kb | +0.00b | 8.33s | +2.68s ⚠️ |
| dist/component.b76c6cb6.js | 22.80kb | +0.00b | 3.27s | -232.00ms 🚀 |
| dist/smartMediaEditor.84449ec5.js | 16.96kb | +0.00b | 13.67s | +751.00ms ⚠️ |
| dist/js.4b5d27df.js | 16.55kb | +0.00b | 3.27s | -232.00ms 🚀 |
| dist/dropzone.f8e6e4ba.js | 16.14kb | +0.00b | 13.59s | +3.90s ⚠️ |
| dist/ui.bb2aade7.js | 14.30kb | +0.00b | 8.80s | +1.61s ⚠️ |
| dist/workerHasher.09e24457.js | 11.90kb | +0.00b | 6.10s | +448.00ms ⚠️ |
| dist/card.4dbff284.js | 5.74kb | +0.00b | 10.39s | +3.18s ⚠️ |
| dist/media-viewer.941270ab.js | 4.01kb | +0.00b | 10.40s | +1.74s ⚠️ |
| dist/EmojiPickerComponent.c8995f9a.js | 3.68kb | +0.00b | 5.62s | -2.48s 🚀 |
| dist/png-chunks-extract.e6351d92.js | 3.55kb | +0.00b | 3.27s | -232.00ms 🚀 |
| dist/index.ece3a13e.css | 3.46kb | +0.00b | 14.45s | -1.92s 🚀 |
| dist/clipboard.dd164e46.js | 3.00kb | +0.00b | 13.60s | +3.95s ⚠️ |
| dist/ResourcedEmojiComponent.fbef555a.js | 2.19kb | +0.00b | 6.40s | +752.00ms ⚠️ |
| dist/card.b6509036.js | 2.15kb | +0.00b | 9.57s | +900.00ms ⚠️ |
| dist/images.fceca81b.js | 1.90kb | +0.00b | 3.89s | -425.00ms 🚀 |
| dist/browser.ea3fd89d.js | 1.84kb | +0.00b | 10.50s | +850.00ms ⚠️ |
| dist/16.934d2c3d.js | 1.79kb | +0.00b | 3.44s | -224.00ms 🚀 |
| dist/workerHasher.852e4f05.js | 1.76kb | +0.00b | 8.32s | +2.67s ⚠️ |
| dist/workerHasher.d451e6c0.js | 1.76kb | +0.00b | 11.22s | -2.04s 🚀 |
| dist/16.d564a7d9.js | 1.46kb | +0.00b | 3.44s | -224.00ms 🚀 |
| dist/16.01c01dd6.js | 1.46kb | +0.00b | 3.45s | -201.00ms 🚀 |
| dist/16.897e8807.js | 1.44kb | +0.00b | 3.77s | -462.00ms 🚀 |
| dist/16.c2dce044.js | 1.40kb | +0.00b | 3.99s | -235.00ms 🚀 |
| dist/16.990be6f2.js | 1.31kb | +0.00b | 3.77s | -461.00ms 🚀 |
| dist/Modal.b44f65e1.js | 1.28kb | +0.00b | 3.27s | -232.00ms 🚀 |
| dist/component.a269da61.js | 1.26kb | +0.00b | 3.44s | +205.00ms ⚠️ |
| dist/16.2ec03e4f.js | 1.26kb | +0.00b | 3.44s | -368.00ms 🚀 |
| dist/16.c20ef86c.js | 1.26kb | +0.00b | 3.44s | -216.00ms 🚀 |
| dist/media-card-analytics-error-boundary.f667fb97.js | 1.25kb | +0.00b | 8.83s | +1.63s ⚠️ |
| dist/16.7b2e1497.js | 1.23kb | +0.00b | 3.44s | +205.00ms ⚠️ |
| dist/panel-warning.02be066d.js | 1.21kb | +0.00b | 4.41s | +310.00ms ⚠️ |
| dist/panel.be955998.js | 1.10kb | +0.00b | 4.44s | +335.00ms ⚠️ |
| dist/panel-success.85b34905.js | 1.05kb | +0.00b | 4.41s | +310.00ms ⚠️ |
| dist/panel-note.dc9835c8.js | 1.05kb | +0.00b | 4.41s | +309.00ms ⚠️ |
| dist/media-picker-analytics-error-boundary.269b86f1.js | 1019.00b | +0.00b | 13.53s | +4.16s ⚠️ |
| dist/media-card-analytics-error-boundary.39871a33.js | 1015.00b | +0.00b | 8.82s | +1.63s ⚠️ |
| dist/media-card-analytics-error-boundary.7ac478f5.js | 1015.00b | +0.00b | 11.22s | +1.56s ⚠️ |
| dist/media-viewer-analytics-error-boundary.bfb50fae.js | 1011.00b | +0.00b | 9.49s | +824.00ms ⚠️ |
| dist/simpleHasher.92ce75a6.js | 763.00b | +0.00b | 11.23s | -2.09s 🚀 |
| dist/index.html | 119.00b | +0.00b | 2.53s | +148.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.511c4e7a.js | 2.46mb | +0.00b | 555.00ms | +55.00ms ⚠️ |
| dist/component.b15bf9ac.js | 31.08kb | +0.00b | 555.00ms | +51.00ms ⚠️ |
| dist/component.b76c6cb6.js | 22.80kb | +0.00b | 479.00ms | -38.00ms 🚀 |
| dist/js.4b5d27df.js | 16.55kb | +0.00b | 479.00ms | -38.00ms 🚀 |
| dist/component.3e8f8d46.js | 6.21kb | +0.00b | 487.00ms | -30.00ms 🚀 |
| dist/png-chunks-extract.e6351d92.js | 3.55kb | +0.00b | 479.00ms | -38.00ms 🚀 |
| dist/Modal.af8c90e8.js | 3.22kb | +0.00b | 467.00ms | -33.00ms 🚀 |
| dist/16.90d86ed2.js | 2.50kb | +0.00b | 467.00ms | -34.00ms 🚀 |
| dist/feedback.26aabd8c.js | 1.87kb | +0.00b | 535.00ms | -29.00ms 🚀 |
| dist/16.81d1aa0c.js | 1.45kb | +0.00b | 467.00ms | -34.00ms 🚀 |
| dist/expand.2e82c578.js | 1.38kb | +0.00b | 535.00ms | -29.00ms 🚀 |
| dist/16.620d8c15.js | 1.34kb | +0.00b | 467.00ms | -34.00ms 🚀 |
| dist/Modal.b44f65e1.js | 1.28kb | +0.00b | 479.00ms | -38.00ms 🚀 |
| dist/16.30ebc170.js | 1.27kb | +0.00b | 467.00ms | -34.00ms 🚀 |
| dist/component.a269da61.js | 1.26kb | +0.00b | 555.00ms | +55.00ms ⚠️ |
| dist/16.7b2e1497.js | 1.23kb | +0.00b | 555.00ms | +55.00ms ⚠️ |
| dist/index.html | 119.00b | +0.00b | 489.00ms | +58.00ms ⚠️ |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 12.42s | +63.00ms |
| Cached | 443.00ms | +2.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 580.60kb | +0.00b | 93.00ms | +6.00ms ⚠️ |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/290336043277746701)**

**Review by mischnic — APPROVED** [review]

File: packages/core/core/src/requests/EntryRequest.js [hunk 1]
```diff
@@ -96,32 +96,46 @@ class EntryResolver {
...
+        let files = [];
+
+        let pkgSources = Array.isArray(pkg.source) ? pkg.source : [pkg.source];
+        for (let pkgSource of pkgSources) {
+          if (typeof pkgSource === 'string') {
+            let source = path.join(path.dirname(pkg.filePath), pkgSource);
+            try {
+              stat = await this.options.inputFS.stat(source);
+            } catch (err) {
+              throw new Error(
```
Comment (devongovett) [inline packages/core/core/src/requests/EntryRequest.js:110]: Should we throw a diagnostic here?

Comment (DeMoorJasper) [reply packages/core/core/src/requests/EntryRequest.js:110]: Probably, will update it

**Comment by mischnic** [comment]
Ideally, there's be a codeframe for the package.json, but the path to that isn't passed into the EntryResolver at the moment

**Comment by DeMoorJasper** [comment]
@mischnic I'll look into improving this further

**Review by devongovett — APPROVED** [review]

**Comment by devongovett** [comment]
you could probably do that separate from this pr if you want

**Comment by DeMoorJasper** [comment]
@devongovett thanks, will refactor it a bit after the source per target is also merged

File: packages/core/core/src/requests/EntryRequest.js [hunk 2]
```diff
@@ -91,40 +92,70 @@ class EntryResolver {
...
+                    this.options.inputFS.cwd(),
+                    pkg.filePath,
+                  )}#source is not a file`,
+                  filePath: source,
+                },
+              });
+            }
+
+            entries.push({filePath: source, packagePath: entry});
+            files.push({filePath: pkg.filePath});
```
Comment (krisnye) [inline packages/core/core/src/requests/EntryRequest.js:140]: @mischnic is this semantically correct? If pkg.source is an array then there will be the same { filePath: pkg.filePath } pushed for every source? seems like only a single one should be pushed.

**Pull request merged**